packages/commonwealth/server/api/external-router-middleware.ts

```diff
@@ -32,48 +38,73 @@ export async function apiKeyAuthMiddleware(
   if (typeof addressHeader !== 'string')
     throw new AppError('Unauthorized', 401);
 
-  const addressInstance = await models.Address.findOne({
-    attributes: ['user_id'],
-    where: {
-      address: addressHeader,
-      verified: { [Op.ne]: null },
-    },
-    include: [
+  let user: z.infer<typeof User> | undefined;
+  const salt = await cache().getKey(
+    CacheNamespaces.Api_key_auth,
+    buildApiKeySaltCacheKey(addressHeader),
+  );
+  if (salt) {
+    const hashedApiKey = getSaltedApiKeyHash(apiKey, salt);
+    const res = await cache().getKey(
+      CacheNamespaces.Api_key_auth,
+      hashedApiKey,
+    );
+    if (res) user = JSON.parse(res);
+  }
+
+  if (!user) {
+    const addressInstance = await models.Address.findOne({
+      attributes: ['user_id'],
+      where: {
+        address: addressHeader,
+        verified: { [Op.ne]: null },
+      },
+      include: [
+        {
+          model: models.User,
+          required: true,
+          include: [
+            {
+              model: models.ApiKey,
+              required: true,
+            },
+          ],
+        },
+      ],
+    });
+    if (!addressInstance || !addressInstance.user_id)
+      throw new AppError('Unauthorized', 401);
+    const address = addressInstance.get({ plain: true })!;
+
+    const apiKeyRecord = address.User!.ApiKey!;
+    const hashedApiKey = getSaltedApiKeyHash(apiKey, apiKeyRecord.salt);
+
+    if (hashedApiKey !== apiKeyRecord.hashed_api_key)
+      throw new AppError('Unauthorized', 401);
+
+    user = address.User!;
+    delete user.ApiKey;
+
+    // cache for 2 minutes
+    await cache().setKeys(
+      CacheNamespaces.Api_key_auth,
       {
-        model: models.User,
-        required: true,
-        include: [
-          {
-            model: models.ApiKey,
-            required: true,
-          },
-        ],
+        [buildApiKeySaltCacheKey(addressHeader)]: apiKeyRecord.salt,
+        [hashedApiKey]: JSON.stringify(user),
       },
-    ],
-  });
-  if (!addressInstance || !addressInstance.user_id)
-    throw new AppError('Unauthorized', 401);
-  const address = addressInstance.get({ plain: true })!;
-
-  const apiKeyRecord = address.User!.ApiKey!;
-  const hashedApiKey = getSaltedApiKeyHash(apiKey, apiKeyRecord.salt);
-
-  if (hashedApiKey !== apiKeyRecord.hashed_api_key)
-    throw new AppError('Unauthorized', 401);
-
-  const user = address.User!;
-  delete user.ApiKey;
+      120,
+    );
+  }
 
   req.user = models.User.build(user);
 
   // record access in background - best effort
-  apiKeyRecord.updated_at = new Date();
   void models.ApiKey.update(
```

Comment:
we are already recording user activity by address... guess this is to track api key activity, but not sure the last api request time helps much... we can probably get much more details from datadog by inspecting headers